server.go

```diff
@@ -218,44 +228,31 @@ func (s *Server) GetUserDataPath(userID string) (string, error) {
 	return filepath.Join(s.dir, userID), nil
 }
 
-func (s *Server) addListener(l net.Listener) {
-	s.listenersLock.Lock()
-	defer s.listenersLock.Unlock()
+// Close closes the server.
+// It firstly closes all TCP listeners then closes the backend.
+func (s *Server) Close(ctx context.Context) error {
+	ctx = reporter.NewContextWithReporter(ctx, s.reporter)
 
-	s.listeners[l] = struct{}{}
+	// Tell the server to stop serving.
+	close(s.serveDoneCh)
 
-	s.publish(events.EventListenerAdded{
-		Addr: l.Addr(),
-	})
-}
-
-func (s *Server) removeListener(l net.Listener) {
-	s.listenersLock.Lock()
-	defer s.listenersLock.Unlock()
+	// Wait until all goroutines currently handling connections are done.
+	s.serveWG.Wait()
 
-	if _, ok := s.listeners[l]; ok {
-		delete(s.listeners, l)
+	// Close the server error channel.
+	close(s.serveErrCh)
```

Comment:
You may want to close this after `s.backend.close()` if a Session close fails for some reason it will result in a panic when trying to write to the err channel.

Reply:
Good point, done